src/rust/engine/process_execution/src/docker_tests.rs

```diff
@@ -66,10 +66,33 @@ macro_rules! setup_docker {
       }
     }
 
-    docker
+    Some(docker)
   }};
 }
 
+#[tokio::test]
+#[cfg(unix)]
+async fn use_local_runner_if_docker_not_set() {
+  // Because Docker is not set up, and docker_image is set, the process should fail.
+  run_command_via_docker(
+    &None,
+    Process::new(owned_string_vec(&["/bin/echo", "-n", "foo"])).docker_image(IMAGE.to_owned()),
+  )
+  .await
+  .unwrap_err();
```

Comment:
AFAICT, the `Option<Docker>` here is only used to pull the image. But afaik, that is only necessary if you don't already have the image locally? So it seems like if you have already pulled the image, then this test might flake?

More generally, it seems like this might be testing a condition which isn't actually implemented. Currently the `docker::CommandRunner` eagerly connects to docker. Because it does that, you cannot have it in the stack and "not" use docker.

To do that, you'd need to have the `CommandRunner` connect lazily on first use. To do that, you could convert the `Docker` field to `async_oncecell::OnceCell<Docker>`, and then initialize it like:
https://github.com/pantsbuild/pants/blob/d2b22d1594803c6a13dfcf6915a02e3b770f91dc/src/rust/engine/process_execution/src/remote.rs#L171-L192

Reply:
> Currently the `docker::CommandRunner` eagerly connects to docker. Because it does that, you cannot have it in the stack and "not" use docker.

Not quite unfortunately. The "connect" in `Docker::connect_with_local_defaults` with the `bollard` crate only sets up the `Docker` struct with the path to the UNIX domain socket, and doesn't actually connect to it until an actual API call is made. This is why I had to add the call to `ping` in the `setup_docker`  macro in `docker_tests.rs`, since I actually needed to verify connectivity.

That said, I agree that lazy initialization will be good practice here because we should not rely on that non-obvious behavior of the `bollard` crate.

Reply:
https://github.com/pantsbuild/pants/issues/16769